recipes/_deps_centos.rb

```diff
@@ -0,0 +1,112 @@
+#
+# Cookbook Name:: mcrouter
+# Recipe:: _build_deps.centos.rb
+#
+# Copyright 2015 EverTrue, Inc.
+# Copyright 2017 Roblox, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+include_recipe 'build-essential'
+include_recipe 'yum'
+include_recipe 'yum-epel'
+
+cache_dir = Chef::Config[:file_cache_path]
+
+%w(
+  centos-release-scl
+  gcc
+  gcc-c++
+  devtoolset-4-gcc
+  devtoolset-4-gcc-c++
+  autoconf
+  binutils-devel
+  bison
+  boost
+  boost-devel
+  double-conversion-devel
+  flex
+  git
+  glog-devel
+  jemalloc-devel
+  libtool
+  libevent-devel
+  make
+  openssl-devel
+  python-devel
+  scons
+  ragel
+  ruby
+  ruby-devel
+  cyrus-sasl
+  cyrus-sasl-devel
+  python-libs
+  python-devel
+  numactl-libs
+  numactl-devel
+  snappy
+  snappy-devel
+  gflags
+  gflags-devel
+  glog
+  automake
+  libevent
+  libatomic
+  rpm-build
+  bzip2-devel
+  libcap-devel
+  unzip
+  glibc-devel.i686
+  glibc-devel.x86_64
+  zlib-devel
+  thrift
+  thrift-devel
+  double-conversion-devel
+).each do |pkg|
+  package pkg
+end
+
+remote_file "#{cache_dir}/libarchive-3.1.2-10.el7_2.x86_64.rpm" do
+  source 'ftp://rpmfind.net/linux/centos/7.3.1611/os/x86_64/Packages/libarchive-3.1.2-10.el7_2.x86_64.rpm'
+end
+
+rpm_package 'libarchive-3.1.2-10.el7_2.x86_64.rpm' do
+  source "#{cache_dir}/libarchive-3.1.2-10.el7_2.x86_64.rpm"
+  not_if "rpm -qa | grep -qx 'libarchive-3.1.2-10.el7_2.x86_6'"
+end
+
+remote_file "#{cache_dir}/automake-1.15-4.fc23.noarch.rpm" do
+  source 'http://dl.fedoraproject.org/pub/fedora/linux/releases/23/Everything/x86_64/os/Packages/a/automake-1.15-4.fc23.noarch.rpm'
+end
+
+rpm_package 'automake-1.15-4.fc23.noarch.rpm' do
+  source "#{cache_dir}/automake-1.15-4.fc23.noarch.rpm"
+  not_if "rpm -qa | grep -qx 'automake-1.15-4.fc23.noarch'"
+end
+
+yum_package 'cmake' do
+  options "--downloadonly --downloaddir=#{cache_dir}"
+end
+
+execute 'force install cmake' do
+  cwd cache_dir
+  command 'rpm -ivh --force cmake-*.rpm'
+  not_if "rpm -qa | grep -qx 'cmake-2.8.12.2-2.el7.x86_64'"
+end
+
+magic_shell_environment 'CC' do
```

Comment:
Have you tried remove just plain gcc/g++? Just guessing here, but the devtools-4 version is needed because something won't compile with GCC version 4 (the default for CentOS) and you need version 5. But since you're overriding `CC` and `CXX` to use the devtools-4 install, it seems the plain gcc/g++ are unnecessary.